File: src/chipschallenge/blockreactions/SocketTo.java

package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.BlockContainerFullException;
import chipschallenge.Game;


public class SocketTo extends BlockReaction {

    private SocketTo() {}
    private static SocketTo mInstance = null;

    public static synchronized SocketTo getInstance() {
        if (mInstance == null) {
            mInstance = new SocketTo();
        }
        return mInstance;
    }

    @Override
    public void react(Block moving, Block standing) throws BlockContainerFullException {
        standing.destroy();
    }

    @Override
    public boolean canMove(Block moving, Block standing) {
        return moving.isChip() && Game.getInstance().getLevel().getNumChipsNeeded() <= 0;
    }

}

File: ChipsChallenge.Shared/Blockreactions/SocketTo.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    public class SocketTo : NoSlipReaction
    {
        private SocketTo()
        {
        }

        private static SocketTo instance;

        public static SocketTo Instance
        {
            get
            {
                lock (typeof(SocketTo))
                {
                    if (instance == null)
                    {
                        instance = new SocketTo();
                    }
                    return instance;
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
            standing.Destroy();
            Sound().Play(Shared.Sound.Socket);
        }

        public override bool canMove(Block moving, Block standing)
        {
            return moving.Chip && Level().NumChipsNeeded <= 0;
        }
    }
}